src/uu/expr/src/syntax_tree.rs

```diff
@@ -197,6 +208,93 @@ const PRECEDENCE: &[&[(&str, BinOp)]] = &[
     &[(":", BinOp::String(StringOp::Match))],
 ];
 
+fn check_brace_content_and_matching(s: &str) -> BraceContent {
```

Comment:
I think this parser is not completely correct. It returns `BraceContent::Invalid` for the following edge cases whereas they seem to work with GNU `expr`:
```
$ expr ab : "\\{\\(\\}\\)"

$ expr ab : "\\(\\{\\)\\}"

$ cargo run -q expr ab : "\\{\\(\\}\\)"
expr: Invalid content of \{\}
$ cargo run -q expr ab : "\\(\\{\\)\\}"
expr: Invalid content of \{\}
```
But I think it's something for a future PR.